include/kernels/multi_reduce_core.cuh

```diff
@@ -13,100 +13,105 @@ namespace quda
   {
 
     /**
-       @brief Parameter struct for generic multi-blas kernel.
-       @tparam NXZ is dimension of input vectors: X,Z,V
-       @tparam NXZ is dimension of input vectors: X,Z
+       @brief Parameter struct for generic multi-reduce blas kernel.
+       @tparam real_ The precision of the calculation
+       @tparam n_ The number of real elements per thread
+       @tparam NXZ_ the dimension of input vectors: X,Z
        @tparam store_t Default store type for the fields
        @tparam N Default field vector i/o length
        @tparam y_store_t Store type for the y fields
-       @tparam N Y-field vector i/o length
-       @tparam Reducer Functor used to operate on data
+       @tparam Ny Y-field vector i/o length
+       @tparam Reducer_ Functor used to operate on data
     */
-    template <int NXZ_, typename store_t, int N, typename y_store_t, int Ny, typename Reducer_>
+    template <typename real_, int n_, int NXZ_, typename store_t, int N, typename y_store_t, int Ny, typename Reducer_>
     struct MultiReduceArg :
       public ReduceArg<vector_type<typename Reducer_::reduce_t, NXZ_>>,
       SpinorXZ<NXZ_, store_t, N, Reducer_::use_z>,
       SpinorYW<max_YW_size<NXZ_, store_t, y_store_t, Reducer_>(), store_t, N, y_store_t, Ny, Reducer_::use_w>
     {
+      using real = real_;
       using Reducer = Reducer_;
+      using reduce_t = vector_type<typename Reducer_::reduce_t, NXZ_>;
+      static constexpr int n = n_;
       static constexpr int NXZ = NXZ_;
       static constexpr int NYW_max = max_YW_size<NXZ, store_t, y_store_t, Reducer>();
       const int NYW;
-      Reducer r;
-      const int length;
-      const int_fastdiv gridSize;
+      Reducer f;
+
+      const int length_cb;
+      const int nParity;
 
       MultiReduceArg(std::vector<ColorSpinorField *> &x, std::vector<ColorSpinorField *> &y,
                      std::vector<ColorSpinorField *> &z, std::vector<ColorSpinorField *> &w,
-                     Reducer r, int NYW, int length, int nParity, TuneParam &tp) :
+                     Reducer f, int NYW, int length, int nParity) :
         // we have NYW * nParity reductions each of length NXZ
-        ReduceArg<vector_type<typename Reducer_::reduce_t, NXZ>>(NYW),
+        ReduceArg<reduce_t>(NYW),
         NYW(NYW),
-        r(r),
-        length(length),
-        gridSize(tp.grid.x * tp.block.x / nParity)
+        f(f),
+        length_cb(length / nParity),
+        nParity(nParity)
       {
+        this->threads = dim3(length, NYW, 1);
```

Comment:
Same comment as `reduce_core.cuh`, this should be propagated via a constructor to the base class, but this has the same base class as in `reduce_core.cuh` so it should be an easy simultaneous fix.

Reply:
1831c77b1